dvc/remote/local/__init__.py

```diff
@@ -569,3 +569,48 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum, force):
+        if self._can_avoid_copy(path_info, checksum):
+            return False
+        return True
+
+    def _is_cache_type_copy(self):
+        if not self.path_info:
+            return False
+
+        try:
+            link_test_dir = self.path_info / "tmp_test"
+            self.makedirs(link_test_dir)
+
+            file1 = link_test_dir / uuid()
+            file2 = link_test_dir / uuid()
+
+            with open(str(file1), "wb") as fobj:
+                fobj.write(bytes(1))
+
+            self.link(file1, file2)
+
+            self.remove(link_test_dir)
+
+            return self.cache_types[0] == "copy"
+        except Exception:
+            return False
+
+    def _is_protected(self, path_info):
+        return not os.access(str(path_info), os.W_OK)
+
+    def _has_same_protection_as_cache(self, path_info):
+        is_protected = self._is_protected(path_info)
+
+        return not bool(self.protected) ^ is_protected
+
+    def _can_avoid_copy(self, path_info, checksum):
+        return (
+            self.exists(path_info)
+            and self.state.get(path_info) == checksum
```

Comment:
I don't think this check is needed. `_checkout_file` is called either from `save` or from `checkout`, both of which are making sure that the file in the workspace has the same checksum.

Reply:
Hm, that is not actually true for `checkout`, sorry.